### PR TITLE
Implementation for <amp-layout>, a generic responsive container extension 

### LINKED ISSUE #10018: Add a generic responsive container component to AMP
### Background
AMP HTML [supports various](https://www.ampproject.org/docs/guides/responsive/control_layout) `layout`s such as aspect-ratio-based responsive layout which is very useful. Other useful layouts such as [responsive-height ](https://github.com/ampproject/amphtml/issues/8531)are also in the works.

### Motivation 
Currently AMP layouts are only supported on AMP elements such a `amp-img`, `amp-carousel`, etc.. but these layouts can be useful on their own.

`amp-layout` component will bring the power of AMP `layout` attribute to a generic multi-purpose container element that can be used independent of other AMP elements.

For instance:
```
<amp-layout class="card" layout=responsive width=2 height=1>
   This card container will always maintain its 2x1 aspect ratio!
</amp-layout>

.card {
  background-color: red;
}
```
### Implementation strategy
- [ ] Create a new `amp-layout` extended component. Only allow `FIXED`, `FIXED_HEIGHT`, `RESPONSIVE`, `FILL`, `FLEX_ITEM` layouts (hint: `isLayoutSizeDefined`)
- [ ] Add validator file for validation rules.
- [ ] Add documentation
- [ ] Add example 

### Step by step
- [ ] Claim this issue by adding a comment below.  Please only claim this bug if you plan on starting work in the next day or so.
- [ ] If you aren't too familiar with Git/GitHub, see the [Getting Started End-to-End Guide](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md) for [an intro to Git & GitHub,](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#intro-to-git-and-github) and [how to get a copy of the code](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#get-a-copy-of-the-amphtml-code).  You can also refer to the [Quick Start Guide](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-quick.md) for the necessary setup steps with less explanation than the End-to-End guide.
- [ ] Follow the instructions for [building AMP](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#building-amp-and-starting-a-local-server).
- [ ] Read up on [How to build an AMP extension](https://docs.google.com/document/d/19o7eDta6oqPGF4RQ17LvZ9CHVQN53whN-mCIeIMM8Qk)
- [ ] [Create a Git branch](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#create-a-git-branch) for making your changes.
- [ ] Follow the **Implementation strategy** from above.
- [ ] [Commit your changes](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#edit-files-and-commit-them) frequently.
- [ ] [Push your changes to GitHub](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#push-your-changes-to-your-github-fork).
- [ ] [Create a Pull Request](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#send-a-pull-request-ie-request-a-code-review).  Mention `closes Issue #10018` in the description.
- [ ] Assign @aghassemi as the reviewer
- [ ] [Respond to your reviewer's comments](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#respond-to-pull-request-comments) (if any).

Once approved, your changes will be merged.  **⚡⚡⚡Congrats on making your first contribution to the AMP Project!⚡⚡⚡**  You'll be able to see it [live across the web soon](https://github.com/ampproject/amphtml/blob/master/contributing/release-schedule.md)!

Thanks, and we hope to see more contributions from you soon.

### Questions?

If you have questions ask in this issue or on your Pull Request (mentioning @aghassemi) or see the [How to get help](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#how-to-get-help) section of the Getting Started guide.

### PR DESCRIPTION
**Pull request by subhankar-panda**
This PR closes #10018

### CODE REVIEW

**Comment by subhankar-panda** [comment]
Not sure how to assign you as a reviewer @aghassemi

File: extensions/amp-layout/validator-amp-layout.protoascii [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    allowed_versions: "latest"
+  }
+  attr_lists: "common-extension-attrs"
+}
+tags: {  # <amp-layout>
+  html_format: AMP
+  tag_name: "AMP-LAYOUT"
+  requires_extension: "amp-layout"
+  attr_lists: "extended-amp-global"
+  attr_lists: "mandatory-src-or-srcset"
```
Comment (honeybadgerdontcare) [inline extensions/amp-layout/validator-amp-layout.protoascii]: The example markdown says nothing is mandatory yet this is including an attribute list for a mandatory src or srcset. Are src or srcset mandatory?

Comment (subhankar-panda) [inline extensions/amp-layout/validator-amp-layout.protoascii]: Oh whoops, that line lingered from when I fixed a merge conflict, my bad - I'll update my branch right away

File: extensions/amp-layout/amp-layout.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# <a name="amp-layout"></a> `amp-layout`
+
+<table>
+  <tr>
+    <td width="40%"><strong>Description</strong></td>
+    <td>A generic, multi-purpose container element that can be used independent of other elements</td>
```
Comment (aghassemi) [inline extensions/amp-layout/amp-layout.md]: `A generic, multi-purpose container element that brings AMP's powerful [layouts](A generic, multi-purpose container element that can be used independent of other elements) such as aspect-ratio based `responsive`, `fixed-height`, etc... to any element.`

/cc @bpaduch

File: extensions/amp-layout/amp-layout.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  </tr>
+  <tr>
+    <td class="col-fourty"><strong><a href="https://www.ampproject.org/docs/guides/responsive/control_layout.html">Supported Layouts</a></strong></td>
+    <td>fill, fixed, fixed-height, flex-item, responsive</td>
+  </tr>
+</table>
+
+## Examples
+
+Use `amp-layout` to create an aspect-ratio based responsive, independent component!.
```
Comment (aghassemi) [inline extensions/amp-layout/amp-layout.md]: Use `amp-layout` to apply aspect-ratio based responsive layouts to any element.

`amp-layout` works similar to the `layout` attribute on existing AMP Extended Components, but supports any HTML markup as children.

[Other layouts](https://www.ampproject.org/docs/guides/responsive/control_layout) such as `fixed-height`, `fixed`, etc... also work on `amp-layout`.

File: package.json [hunk 4]
```diff
@@ -52,7 +53,7 @@
     "formidable": "1.0.17",
     "fs-extra": "0.27.0",
     "glob": "7.1.2",
-    "gulp": "3.9.1",
+    "gulp": "^3.9.1",
```
Comment (aghassemi) [inline package.json]: please revert this and the `lock` file changes.

File: extensions/amp-layout/amp-layout.md (hunk 2 above)
Comment (bpaduch) [inline extensions/amp-layout/amp-layout.md]: For my understanding, the purpose of this component is to use 'layout=xxx' for non-AMP HTML elements?

What would be the obvious, most common use case for this component?

File: extensions/amp-layout/amp-layout.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+**Example**
+
+```html
+<amp-layout class="card" layout=responsive width=2 height=1>
```
Comment (bpaduch) [inline extensions/amp-layout/amp-layout.md]: For my edification, should code examples have attribute values in quotes or not?  I know it's a no-op, but stylistically, what's preferred and clear?

File: extensions/amp-layout/amp-layout.md (hunk 2 above)
Comment (aghassemi) [inline extensions/amp-layout/amp-layout.md]: Yes. I would say a common use-case is image-like widgets where instead of using an actual image, they are drawn by CSS or inline SVG.

File: extensions/amp-layout/amp-layout.md (hunk 5 above)
Comment (aghassemi) [inline extensions/amp-layout/amp-layout.md]: good catch, with quotes is preferred @subhankar-panda

File: extensions/amp-layout/amp-layout.md (hunk 2 above)
Comment (bpaduch) [inline extensions/amp-layout/amp-layout.md]: Using aghassemi's words + conciseness, please rewrite the short description as follows:

`A generic, multi-purpose container element that brings AMP's powerful [layouts](https://www.ampproject.org/docs/guides/responsive/control_layout#the-layout-attribute) to any element.`

File: extensions/amp-layout/amp-layout.md (hunk 3 above)
Comment (bpaduch) [inline extensions/amp-layout/amp-layout.md]: Using aghassemi's words + conciseness + consistency, please create an overview section and include the example in that section like the following:

## Overview

The `amp-layout` component allows you to apply aspect-ratio based responsive layouts to any element. The `amp-layout` component works similarly to the [layout](https://www.ampproject.org/docs/guides/responsive/control_layout#the-layout-attribute) attribute on existing AMP components, but supports any HTML markup as children. Other supported layouts all work with `amp-layout` (e.g., fixed-height, fixed, etc.). 

### Example

< code sample here >

File: extensions/amp-layout/amp-layout.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  This card container will always maintain its 2x1 aspect ratio!
+</amp-layout>
+
+.card {
```
Comment (bpaduch) [inline extensions/amp-layout/amp-layout.md]: Should this CSS be in the code sample?  Typically, CSS is in `<style amp-custom`> in the head ... it seems confusing to have it here, and to me adds no value.

It'd be nice to see an example that @aghassemi mentioned as a common use case.

File: extensions/amp-layout/amp-layout.md (hunk 3 above)
Comment (bpaduch) [inline extensions/amp-layout/amp-layout.md]: Using aghassemi's words + conciseness + consistency, please create an overview section and include the example in that section like the following:

## Overview

The `amp-layout` component allows you to apply aspect-ratio based responsive layouts to any element. The `amp-layout` component works similarly to the [layout](https://www.ampproject.org/docs/guides/responsive/control_layout#the-layout-attribute) attribute on existing AMP components, but supports any HTML markup as children. Other supported layouts all work with `amp-layout` (e.g., fixed-height, fixed, etc.). 

### Example

< code sample here >

File: extensions/amp-layout/amp-layout.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+-->
+
+# <a name="amp-layout"></a> `amp-layout`
+
+<table>
+  <tr>
+    <td width="40%"><strong>Description</strong></td>
+    <td>A generic, multi-purpose container element that can be used independent of other elements</td>
+  </tr>
+  <tr>
```
Comment (bpaduch) [inline extensions/amp-layout/amp-layout.md]: BTW, is this component "stable", that is, already in production? Or experimental?  If not in production, please add an "availability" row with the availability status (similar to what [amp-viz-vega](https://github.com/ampproject/amphtml/blob/master/extensions/amp-viz-vega/amp-viz-vega.md) has): 
```html
  <tr>
    <td width="40%"><strong>Availability</strong></td>
    <td><a href="https://www.ampproject.org/docs/reference/experimental.html">Experimental</a></td>
  </tr>
```

**Comment by aghassemi** [comment]
@subhankar-panda just wanted to follow up on requested changes for this PR.

File: extensions/amp-layout/amp-layout.md (hunk 6 above)
Comment (subhankar-panda) [inline extensions/amp-layout/amp-layout.md]: I added an example that uses amp-layout around an inline SVG definition but I'm not sure if it's very helpful, please have a look

Comment (bpaduch) [inline extensions/amp-layout/amp-layout.md]: I like the example.  Makes sense.

File: package.json [hunk 8]
```diff
@@ -86,7 +86,7 @@
     "karma-mocha": "0.2.2",
     "karma-safari-launcher": "0.1.1",
     "karma-sauce-launcher": "1.1.0",
-    "karma-sinon-chai": "1.2.0",
+    "karma-sinon-chai": "1.2.3",
```
Comment (aghassemi) [inline package.json]: @subhankar-panda please revert changes to `package.json`, `package-lock.json` and `yarn.lock`, otherwise looks good and is ready to merge after.

**Comment by aghassemi** [comment]
@subhankar-panda friendly ping to rebase so we can merge this PR

File: package.json [hunk 9]
```diff
@@ -123,3 +123,4 @@
     "failFast": true
   }
 }
+
```
Comment (rsimha-amp) [inline package.json]: Revert this change and the addition of the lock file.

**Comment by rsimha-amp** [comment]
package.json* changes approved

**Comment by aghassemi** [comment]
Thanks @subhankar-panda 
/cc @dvoytenko  @ericlindley-g: Planning to merge this, please let me know if any concerns.

File: extensions/amp-layout/0.1/amp-layout.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+class AmpLayout extends AMP.BaseElement {
+
+    /** @override */
+    isLayoutSupported(layout){
+        return isLayoutSizeDefined(layout);
+    }   
+}
+
+AMP.registerElement('amp-layout', AMPLayout);
```
Comment (dvoytenko) [inline extensions/amp-layout/0.1/amp-layout.js]: It looks like the classname is wrong here - wrong capitalization.

**Comment by mrjoro** [comment]
Hi @subhankar-panda; are you still working on this PR?

**Comment by jridgewell** [comment]
This has been updated, please rereview @dvoytenko and @aghassemi.

**Comment by aghassemi** [comment]
Thanks @subhankar-panda

**Comment by aghassemi** [comment]
Merging, will send PR to move to builtin as discussed offline with @dvoytenko

**Comment by Gregable** [comment]
Validator changes are live everywhere.
